### PR TITLE
Crop Screenshot to fit the Bounds of a WindowEvent (#108

### PR DESCRIPTION
**Pull request by AvidEslami**
Solves #84 
Summary: Added screenshot.crop_active_window,
- Crops screenshot._image using the bounds of the window_event input parameter.
- Tested using WindowEvents from Recording:
``` Python
#sample test code used in DemoReplayStrategy
screenshot._image.show()  #image 1
screenshot.crop_active_window(window_event)
screenshot._image.show()  #image 2
```

Test Result:
**Image 1:**
image
**Image 2:**
image

TODO: Find the WindowEvent of a screenshot that was not part of a recording. This might be done through a helper function: screenshot.get_window_event which could be called from within crop_active_window instead of passing the WindowEvent as an input parameter.

### CODE REVIEW

**Comment by abrichr** [comment]
Thanks @AvidEslami ! Can you please also show some usage example, ideally as code, or if not as documentation?

**Comment by abrichr** [comment]
@AvidEslami any update on this?

**Comment by AvidEslami** [comment]
> Thanks @AvidEslami ! Can you please also show some usage example, ideally as code, or if not as documentation?

Hi! I created a temporary file to function as a quick test, run the python file as follows:
```Python
python .\openadapt\temp_crop_test.py
```
This test will take a screenshot of your current screen, retrieve the first WindowEvent from your latest recording, and will display the cropped and uncropped versions of that screenshot relative to the bounds of the WindowEvent. 

__Status Summary:__
Currently you can crop the screenshots with reference to a WindowEvent if you have access to one, but finding the WindowEvent for a screenshot isn't the easiest unless it is from a recording. In the [Window state pull request](https://github.com/MLDSAI/OpenAdapt/pull/151/files#:~:text=%22screenshots%22-,action_event,-%3D%20sa.) there are some additions to the screenshot class which could make it easier to find the WindowEvents. I'll look into these new changes to see if improvements/additions could be made.

Let me know if there are any issues or suggestions/improvements to be made!

**Comment by abrichr** [comment]
Thanks @AvidEslami ! Those changes have been merged to master so you should be good to go 👍

**Comment by 0dm** [comment]
got it working, looks good 👍

**Comment by 0dm** [comment]
thought it was working, but looks like it's not cropping properly because it's missing ratios:


solution: pass action_event instead of window_event and use this:
```python
    
    def crop_active_window(self, action_event):
        window_event = action_event.window_event
        
        width_ratio, height_ratio = utils.get_scale_ratios(action_event)
        
        x0 = window_event.left * width_ratio
        y0 = window_event.top * height_ratio
        x1 = x0 + window_event.width * width_ratio
        y1 = y0 + window_event.height * height_ratio
        box = (x0, y0, x1, y1)

        self._image = self._image.crop(box)
  ```
  before:
  recording-1-0

  after:
  recording-1-0

File: openadapt/models.py [hunk 1]
```diff
@@ -273,6 +273,28 @@ def take_screenshot(cls):
...
+            box = (
+                accessed_window_event.left,
+                accessed_window_event.top,
+                accessed_window_event.left + accessed_window_event.width,
+                accessed_window_event.top + accessed_window_event.height
+            )
+        else:
+            logger.info("No window_event found or given for screenshot")  
+            return
+        self._image = self._image.crop(box)
```
Comment (0dm) [inline openadapt/models.py:297]:
```suggestion
    def crop_active_window(self, action_event):
        window_event = action_event.window_event
        
        width_ratio, height_ratio = utils.get_scale_ratios(action_event)
        
        x0 = window_event.left * width_ratio
        y0 = window_event.top * height_ratio
        x1 = x0 + window_event.width * width_ratio
        y1 = y0 + window_event.height * height_ratio
        box = (x0, y0, x1, y1)

        self._image = self._image.crop(box)
```

File: openadapt/temp_crop_test.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import fire
+from openadapt.crud import get_latest_recording, get_window_events
+from openadapt.utils import configure_logging, get_strategy_class_by_name
+from openadapt.models import Screenshot
+
+
+def crop_test():
+    recording = get_latest_recording()
+    window_event = get_window_events(recording)[0]
```
Comment (0dm) [inline openadapt/temp_crop_test.py:9]: use actionevent instead then, for ratios

**Review by 0dm — CHANGES_REQUESTED** [review]

File: openadapt/models.py [hunk 3]
```diff
@@ -274,26 +274,24 @@ def take_screenshot(cls):
...
-            )
-        elif self.action_event.window_event:
-            accessed_window_event = self.action_event.window_event
-            box = (
-                accessed_window_event.left,
-                accessed_window_event.top,
-                accessed_window_event.left + accessed_window_event.width,
-                accessed_window_event.top + accessed_window_event.height
-            )
+    def crop_active_window(self, action_event=None):
```
Comment (0dm) [inline openadapt/models.py:277]: > **Warning**
Make action_event a required parameter

input (within visualize.py) :
```python
        im = action_event.screenshot
        im.crop_active_window() # => im.crop_active_window(action_event)
```
output:
```python
    window_event = action_event_to_use.window_event
AttributeError: 'InstrumentedList' object has no attribute 'window_event'
```

**Review by 0dm — CHANGES_REQUESTED** [review]

File: openadapt/temp_crop_test.py [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+from openadapt.models import Screenshot
+
+
+def crop_test():
+    recording = get_latest_recording()
+    action_event = recording.action_events[0]
+    screenshot = Screenshot.take_screenshot()
+    screenshot.image.show()
+    screenshot.crop_active_window(action_event)
+    screenshot.image.show()
```
Comment (abrichr) [inline openadapt/temp_crop_test.py:13]: Please assert something about the output vs. expected

**Pull request merged**